ocaml/xapi/daemon_manager.ml

```diff
@@ -0,0 +1,90 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Threadext
+
+module IntSet = Set.Make(struct type t = int let compare = compare end)
+
+type daemon_check =
+	| Pidfile of string
+	| Function of (unit -> bool)
+
+type daemon_state = [
+	`unmanaged |
+	`should_start |
+	`should_not_start
+]
+
+module type DAEMON = sig
+	val check : daemon_check
+
+	val start : unit -> unit
+
+	val stop : unit -> unit
+end
+
+module Make(D : DAEMON) = struct
+	let registered_threads = ref IntSet.empty
+
+	let register_thread id =
+		let registered = !registered_threads in
+		registered_threads := (IntSet.add id registered)
+
+	let deregister_thread id =
+		let registered = !registered_threads in
+		registered_threads := (IntSet.remove id registered)
+
+	let are_threads_registered () =
```

Comment:
Can this and the above two functions be appended with `_nolock` or equivalent to make it clear that they should be protected by a mutex. I can see they are being protected later in the module and that they are not exposed in the `.mli` but it's nice to be explicit about these things if you have no objection.